File: backend/src/main/scala/com/skewwhiffy/auraltester/abc/NoteFactory.scala
package com.skewwhiffy.auraltester.abc

import com.skewwhiffy.auraltester.notes.{AbsoluteNote, Accidental, Note, Octave}
import scala.util.chaining._

import scala.annotation.tailrec

class NoteFactory(private val abc: String) {
  lazy val absoluteNote: AbsoluteNote = AbsoluteNote(note, octave)

  private lazy val note: Note = Note(noteName, accidental)

  private lazy val noteName: String = {
    rawNote.toUpperCase match
      case it if it >= "A" && it <= "G" => it
      case it => throw IllegalArgumentException(s"'$it' is not a valid note name'")
  }

  private lazy val accidental: Accidental = rawAccidental
    .toCharArray
    .map(it => it match
      case 'x' => 2
      case '#' => 1
      case 'b' => -1
      case _ => throw IllegalArgumentException(s"Not valid accidental: '$it'")
    )
    .sum
    .pipe(it => Accidental(it))

  private val octave: Octave = rawOctave
    .toCharArray
    .map(it => it match
      case '\'' => 1
      case ',' => -1
      case _ => throw IllegalArgumentException(s"Not valid octave indicator: '$it'")
    )
    .sum
    .pipe(it => it + (if rawNote.toLowerCase == rawNote then 1 else 0))
    .pipe(it => Octave(it))

  private lazy val rawNote: String = abc.substring(0, 1)

  private lazy val rawAccidental: String = {
    @tailrec
    def getRawAccidental(str: String): String = {
      if str.endsWith("'") || str.endsWith(",") then getRawAccidental(str.substring(0, str.length - 1))
      else str
    }

    getRawAccidental(abc.substring(1))
  }

  private lazy val rawOctave: String = abc
    .substring(rawNote.length + rawAccidental.length)
}

File: backend.old/src/main/kotlin/com/skewwhiffy/auraltester/notation/factory/NoteFactory.kt
package com.skewwhiffy.auraltester.notation.factory

import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote
import com.skewwhiffy.auraltester.notation.model.note.Accidental
import com.skewwhiffy.auraltester.notation.model.note.Note
import com.skewwhiffy.auraltester.notation.model.note.Octave
import org.springframework.stereotype.Service

import java.lang.IllegalArgumentException

@Service
class NoteFactory {
    fun getAbsoluteNote(rawNote: String) = AbsoluteNote(getNote(rawNote), getOctave(rawNote), null)

    fun getNote(rawNote: String) = Note(getNoteName(rawNote), getAccidental(rawNote))

    private fun getNoteName(rawNote: String) = getNoteLetter(rawNote)
        .uppercase()
        .apply { if (this < "A" || this > "G") throw throw IllegalArgumentException("'$this' is not a valid note name") }

    private fun getAccidental(rawNote: String): Accidental {
        val getOffset: (Char) -> Int = {
            when (it) {
                'x' -> 2
                '#' -> 1
                'b' -> -1
                else -> throw IllegalArgumentException("Not valid accidental: '$it'")
            }
        }
        return getRawAccidental(rawNote).sumOf(getOffset).let(::Accidental)
    }

    private fun getOctave(rawNote: String): Octave {
        val getOffset: (Char) -> Int = {
            when (it) {
                '\'' -> 1
                ',' -> -1
                else -> throw IllegalArgumentException("Not valid octave indicator: '$it'")
            }
        }
        val offsetBase = getRawOctave(rawNote).sumOf { getOffset(it) }
        val offset = offsetBase + if (rawNote.lowercase() == rawNote) 1 else 0
        return Octave(offset)
    }


    private fun getNoteLetter(rawNote: String) = rawNote.substring(0, 1)

    private fun getRawAccidental(rawNote: String) = rawNote
        .substring(1)
        .let(::getRawAccidentalFromAccidentalMarks)

    private fun getRawAccidentalFromAccidentalMarks(accidentalMarks: String): String =
        if (accidentalMarks.endsWith("'") || accidentalMarks.endsWith(","))
            getRawAccidentalFromAccidentalMarks(accidentalMarks.substring(0, accidentalMarks.length - 1))
        else accidentalMarks

    private fun getRawOctave(rawNote: String) = rawNote
        .substring(getNoteLetter(rawNote).length + getRawAccidental(rawNote).length)
}
